### PR TITLE
PMM-9947: Encryption

### PR DESCRIPTION
**Pull request by ritbl**
PMM-9947

E2E encryption proposal.

FB:
- https://github.com/Percona-Lab/pmm-submodules/pull/2552

### CODE REVIEW

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Thanks, finally I found a time to review this PR. Sorry for that.
How are we going to share public and private keys between pmm-admin and pmm-agent? Are we going to inject them during the build of PMM or just store them in repository or generate them on PMM managed side and share accross clients? 
If we will store them in repository than still it's not secure. pmm-admin allows to add agents to another host and if pmm-agent on another host doesn't support ecryption we will have a problem.
If we will inject keys during build we should disable encryption by default for builds made by community to not have problems with compatability.

Tests are failing. Please fix them.
Please fix linters as well.

File: admin/commands/management/add_mongodb.go [hunk 1]
```diff
@@ -150,6 +157,13 @@ func (cmd *addMongoDBCommand) Run() (commands.Result, error) {
 		}
 	}
 
+	encryptor := rsa_encryptor.GetEncryptor(ctx)
```
[the remark this answers is not in the document]
Comment (ritbl) [reply admin/commands/management/add_mongodb.go:160]: I'll fix the tests

File: utils/rsa_encryptor/encryptor.go [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+	"encoding/pem"
+	"io/fs"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (ritbl) [reply utils/rsa_encryptor/encryptor.go:21]: Ok. Will make them private

File: utils/rsa_encryptor/encryptor.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+	EncryptedTextBlockParamsDelimiter = ":"
+	EncryptedTextBlockCipherStart     = "["
+	EncryptedTextBlockCipherEnd       = "]"
+)
+
+// Formatter formats given string.
+type Formatter func(string) string
+
+// Service provides RSA encryption interface for sensitive data.
+type Service struct {
```
[the remark this answers is not in the document]
Comment (ritbl) [reply utils/rsa_encryptor/encryptor.go:32]: >can we call it Encryptor?
ok

>Too many public methods, let's make methods private if we don't use them outside of this package.
What method/methods you think should be made private?

**Comment by ritbl** [comment]
@BupycHuk 

>How are we going to share public and private keys between pmm-admin and pmm-agent?
We have multiple options:
1: Leave as it is in this PR - anyone with a bit of patience can find all the keys and decrypt data 
Security: 1 out of 5 (very bad). Just grab the keys from repo....
Ease of change / backward compatibility: 5 out of 5 (great).

2: Inject keys in the build time. Someone needs to generate keys and put in secrets in the repo.
Security: 2 out of 5 (bad). Keys will be embedded. It will make harder to get the key but it is still very easy.
Ease of change / backward compatibility: 4 out of 5 (good)

3: Inject keys before pmm starts. 
Security: 3-4 out of 5 (good). Proper solution will take time and client now need to manage the keys. It can be make semi-transparent, but still it will require a lot more work.
Ease of change / backward compatibility: 1 out of 5 (very bad)

We can start with any of these, the key can be rotated later. I think any solution would be better if we compare it with the current one.


>by default for builds made by community to not have problems with compatability
Actually, it is not needed. It will all work. It is backward compatible. We can start with key being even stored in source code (just because it is easy). Later we can provide better key management solution and create simple rotation job. The key name is embedded into metadata, so that we can have multiple keys.

**Comment by BupycHuk** [comment]
> pmm-admin allows to add agents to another host; if pmm-agent on another host doesn't support encryption, we will have a problem.
it's not a typical case, but it's still possible.

File: utils/rsa_encryptor/encryptor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (BupycHuk) [reply utils/rsa_encryptor/encryptor.go]: I think I was clear here. All methods are not used outside of this package.

**Comment by ritbl** [comment]
@BupycHuk 

>pmm-admin allows to add agents to another host; if pmm-agent on another host doesn't support encryption, we will have a problem.
>it's not a typical case, but it's still possible.

Admin perform encryption, for that public key is used. It doesn't need to be kept in secret.
We can embed public key and even put in the source code. Only private key should be taken care of (agent).

**Review by atymchuk — COMMENTED** [review]
@ritbl Do you mind if I convert it to draft to suppress notifications?

**Pull request closed without merging**